lua/lazyvim/plugins/extras/lang/python.lua

```diff
@@ -0,0 +1,116 @@
+return {
+  {
+    "nvim-treesitter/nvim-treesitter",
+    opts = function(_, opts)
+      if type(opts.ensure_installed) == "table" then
+        vim.list_extend(opts.ensure_installed, { "ninja", "python", "rst", "toml" })
+      end
+    end,
+  },
+  {
+    "jose-elias-alvarez/null-ls.nvim",
+    opts = function(_, opts)
+      local nls = require("null-ls")
+      opts.sources = opts.sources or {}
+      vim.list_extend(opts.sources, {
+        -- Order of formatters matters. They are used in order of appearance.
+        nls.builtins.formatting.ruff.with({ filetypes = { "python" }, prefer_local = ".venv/bin" }),
+        nls.builtins.formatting.black.with({ filetypes = { "python" }, prefer_local = ".venv/bin" }),
+      })
+    end,
+    dependencies = {
+      "mason.nvim",
+      opts = function(_, opts)
+        opts.ensure_installed = opts.ensure_installed or {}
+        vim.list_extend(opts.ensure_installed, { "black", "ruff" })
+      end,
+    },
+  },
+  {
+    "neovim/nvim-lspconfig",
+    opts = {
+      servers = {
+        pyright = {
+          settings = {
+            pyright = {
+              disableLanguageServices = false,
+              disableOrganizeImports = false,
+            },
+            python = {
+              analysis = {
+                autoImportCompletions = true,
+                autoSearchPaths = true,
+                diagnosticMode = "workspace",
+                typeCheckingMode = "basic",
+                useLibraryCodeForTypes = true,
+              },
+            },
+          },
+        },
+        ruff_lsp = {
+          settings = {
+            ruff_lsp = {
+              init_options = {
+                settings = {
+                  args = {},
+                },
+              },
+            },
+          },
+        },
+      },
+    },
+    setup = {
+      ruff_lsp = function()
+        require("lazyvim.util").on_attach(function(client, _)
+          if client.name == "ruff_lsp" then
+            -- Disable hover in favor of Pyright
+            client.server_capabilities.hoverProvider = false
+          end
+        end)
+      end,
+    },
+  },
+  {
+    "nvim-neotest/neotest",
+    optional = true,
+    dependencies = {
+      "nvim-neotest/neotest-python",
+    },
+    opts = {
+      adapters = {
+        ["neotest-python"] = {
+          -- Extra arguments for nvim-dap configuration
+          -- See https://github.com/microsoft/debugpy/wiki/Debug-configuration-settings for values
+          dap = {
+            justMyCode = false,
+            console = "integratedTerminal",
+          },
+          args = { "--log-level", "DEBUG", "--quiet" },
+          runner = "pytest",
+          python = ".venv/bin/python",
```

Comment:
I'm not sure, but it might be better to take the defaults for these, and by setting `runner` and `python`, I think you lose the "dynamic" fallback behavior.

https://github.com/nvim-neotest/neotest-python:
>         -- Runner to use. Will use pytest if available by default.
>         -- Can be a function to return dynamic value.
>         runner = "pytest",
>         -- Custom python path for the runner.
>         -- Can be a string or a list of strings.
>         -- Can also be a function to return dynamic value.
>         -- If not provided, the path will be inferred by checking for 
>         -- virtual envs in the local directory and for Pipenev/Poetry configs
>         python = ".venv/bin/python",

Reply:
Yea you are right here. I changed the code to use default options to the neotest-python adapter.